analyzers/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs

```diff
@@ -140,7 +140,10 @@ private static void CheckTypeName(SonarSyntaxNodeReportingContext context)
             }
 
             if (symbol.DeclaringSyntaxReferences.Length > 1
-                && symbol.DeclaringSyntaxReferences.Any(syntax => syntax.SyntaxTree.IsConsideredGenerated(CSharpGeneratedCodeRecognizer.Instance, context.Compilation)))
+                && symbol.DeclaringSyntaxReferences.Any(syntax => syntax.SyntaxTree.IsConsideredGenerated(
+                       CSharpGeneratedCodeRecognizer.Instance,
+                       context.Compilation,
+                       context.SonarLintXml().AnalyzeRazorCode(context.Compilation.Language))))
```

Comment:
I've noticed that:
- `context.SonarLintXml().AnalyzeRazorCode(context.Compilation.Language))))` is called quite a lot and it might be expensive (possible IO operation)
- `SonarLintXml` is defined in the `SonarAnalysisContextBase` where the compilation is already known, which means it will not outlive the compilation
- Could we add a property (with a lazy backing field) on the context `IsRazorAnalysisEnabled` that will cache the result?

It is also possible that this is not needed, it would be great to measure the impact of calling it.

Reply:
The `SonarAnalysisContextBase.SonarLintXml` already does the caching where the `SonarLintXml SourceText` is the key. So, for 1 SL.xml it will create 1 reader which will in the constructor parse the file and after every other use just return the already parsed value. And since it is the same caching mechanism as for the other parameters (and we never had problems with it) I suggest merging as is and check on peach if there is any perf regression.